internal/backup/csi_volumesnapshot.go

```diff
@@ -0,0 +1,120 @@
+func (p *VolumeSnapshotBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotBackupItemAction AppliesTo")
+
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshots.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+func setVolumeSnapshotAnnotationsAndLabels(vs *snapshotv1beta1api.VolumeSnapshot, vals map[string]string) {
+	if vs.Annotations == nil {
+		vs.Annotations = make(map[string]string)
+	}
+	if vs.Labels == nil {
+		vs.Labels = make(map[string]string)
+	}
+
+	for k, v := range vals {
+		vs.Annotations[k] = v
+		vs.Labels[k] = v
+	}
+}
+
+// Execute backs VolumeSnapshotContents associated with the volumesnapshot as additional resource.
+func (p *VolumeSnapshotBackupItemAction) Execute(item runtime.Unstructured, backup *velerov1api.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+	p.Log.Infof("Executing VolumeSnapshotBackupItemAction")
+
+	var vs snapshotv1beta1api.VolumeSnapshot
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &vs); err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	_, snapshotClient, err := util.GetClients()
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+	p.Log.Infof("Getting VolumesnapshotContent for Volumesnapshot %s/%s", vs.Namespace, vs.Name)
+	vsc, err := util.GetVolumeSnapshotContentForVolumeSnapshot(&vs, snapshotClient.SnapshotV1beta1(), p.Log)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	vals := map[string]string{
+		util.VolumeSnapshotHandleLabel: *vsc.Status.SnapshotHandle,
+		util.CSIDriverNameLabel:        vsc.Spec.Driver,
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+
+	// Capture the vsc's deletesnapshot secret annotations into the backup of the volumesnapshot
+	if util.IsVolumeSnapshotContentHasDeleteSecret(vsc) {
+		deleteSnapshotSecretName := vsc.Annotations[util.PrefixedSnapshotterSecretNameKey]
+		deleteSnapshotSecretNamespace := vsc.Annotations[util.PrefixedSnapshotterSecretNamespaceKey]
+		// TODO: add GroupResource for secret into kuberesource
+		additionalItems = append(additionalItems, velero.ResourceIdentifier{
+			GroupResource: schema.GroupResource{Group: "", Resource: "secrets"},
+			Name:          deleteSnapshotSecretName,
+			Namespace:     deleteSnapshotSecretNamespace,
+		})
+
+		vals[util.CSIDeleteSnapshotSecretName] = deleteSnapshotSecretName
+		vals[util.CSIDeleteSnapshotSecretNamespace] = deleteSnapshotSecretNamespace
+
+		p.Log.Infof("Found DeleteSnapshotSecret %s/%s on volumesnapshotcontent %s from its annotation", deleteSnapshotSecretNamespace, deleteSnapshotSecretName, vsc.Name)
+	} else {
+		p.Log.Infof("Volumesnapshotcontent %s does not have a deletesnapshot secret annotation", vsc.Name)
+	}
+
+	// save newly applied annotations and labels into the backed-up volumesnapshot
+	setVolumeSnapshotAnnotationsAndLabels(&vs, vals)
```

Comment:
This code copies the secret info off of the Class & Content objects and puts it on the VolumeSnapshot... when we restore, will this get copied from the snapshot back to the appropriate places?

The VolumeSnapshotClass should transfer over unchanged (at least, I think), but the VolumeSnapshotContents object will need to be recreated and the secret annotation added before the VolumeSnapshot is made, so either we need to read ahead and get the secret annotation off the VolumeSnapshot in the backup, or we need to store it another way, I think.

If https://github.com/vmware-tanzu/velero/pull/2323 is in, then we have all the VolumeSnapshots and VolumeSnapshotContents items stored in object storage as JSON arrays that could be downloaded without unpacking the whole tarball. That may help with order of operations.